sound/soc/sof/virtio-be.c

```diff
@@ -5,122 +5,1039 @@
+#include <linux/miscdevice.h>
 #include <linux/fs.h>
+#include <linux/file.h>
+#include <linux/hw_random.h>
+#include <linux/uio.h>
 #include <linux/slab.h>
+#include <linux/vmalloc.h>
 #include <linux/device.h>
-#include <linux/interrupt.h>
-#include <linux/module.h>
-#include <linux/dma-mapping.h>
-#include <linux/platform_device.h>
-#include <linux/firmware.h>
-#include <virtio.h>
-#include <uapi/sound/sof-fw.h>
+#include <sound/sof.h>
+#include <sound/pcm_params.h>
+#include <uapi/sound/sof-ipc.h>
+#include <linux/vbs/vq.h>
+#include <linux/vbs/vbs.h>
+#include <linux/vhm/acrn_common.h>
+#include <linux/vhm/acrn_vhm_ioreq.h>
+#include <linux/vhm/acrn_vhm_mm.h>
+#include <linux/vhm/vhm_vm_mngt.h>
 #include "sof-priv.h"
 #include "ops.h"
+#include "virtio-be.h"
+#include "virtio-miscdev.h"
 
-/* BE driver
- *
- * This driver will create IO Queues for communition from FE drivers.
- * The FE driver will send real IPC structures over the queue and then
- * the BE driver will send the structures directlt to the DSP. The BE will
- * get the IPC reply from the DSP and send it back to the FE over the queue.
- *
- * The virt IO message Q handlers in this file will :-
- *
- * 1) Check that the message is valid and not for any componenets that don't
- *    belong to the guest.
- *
- * 2) Call snd_sof_dsp_tx_msg(struct snd_sof_dev *sdev,
- *	struct snd_sof_ipc_msg *msg) to send the message to the DSP.
- *
- * Replies will be sent back using a similar method.
+/* TODO: move the below macros to the headers */
+#define IGS(x) (((x) >> SOF_GLB_TYPE_SHIFT) & 0xf)
+#define ICS(x) (((x) >> SOF_CMD_TYPE_SHIFT) & 0xfff)
+
+/* find client from client ID */
+static struct sof_vbe_client *vbe_client_find(struct snd_sof_dev *sdev,
+					      int client_id)
+{
+	struct sof_vbe_client *client;
+	struct sof_vbe *vbe;
+
+	list_for_each_entry(vbe, &sdev->vbe_list, list) {
+		list_for_each_entry(client, &vbe->client_list, list) {
+			if (client_id == client->vhm_client_id)
+				return client;
+		}
+	}
+
+	return NULL;
+}
+
+static struct sof_vbe *sbe_comp_id_to_vbe(struct snd_sof_dev *sdev, int comp_id)
+{
+	struct sof_vbe *vbe;
+
+	list_for_each_entry(vbe, &sdev->vbe_list, list) {
+		if (comp_id < vbe->comp_id_end && comp_id >= vbe->comp_id_begin)
+			return vbe;
+	}
+	return NULL;
+}
+
+static int sof_virtio_send_ipc(struct snd_sof_dev *sdev, void *ipc_data,
+			       void *reply_data, size_t count,
+			       size_t reply_size)
+{
+	struct snd_sof_ipc *ipc = sdev->ipc;
+	struct sof_ipc_hdr *hdr = (struct sof_ipc_hdr *)ipc_data;
+
+	return sof_ipc_tx_message(ipc, hdr->cmd, ipc_data, count,
+				  reply_data, reply_size);
+}
+
+/* Return true if posn buffer is filled successfully */
+static bool sbe_fill_posn_vqbuf(struct sof_vbe *vbe,
+				struct virtio_vq_info *vq,
+				struct sof_ipc_stream_posn *posn,
+				bool *endchain)
+{
+	struct device *dev = vbe->sdev->dev;
+	struct iovec iov;
+	u16 idx;
+	int ret;
+
+	*endchain = false;
+	while (virtio_vq_has_descs(vq)) {
```

Comment:
please comment what we are doing in here and the decisions we make..

Reply:
sure, I will add more comments here and the below you mentioned and other places where the code is complicated.